src/tarball.rs

```diff
@@ -0,0 +1,166 @@
+//! Pin a tarball URL source
+//!
+//! Optionally (if the host supports it) can use the "Lockable HTTP Tarball Protocol" from flakes.
+//! Reference: <https://github.com/nixos/nix/blob/56763ff918eb308db23080e560ed2ea3e00c80a7/doc/manual/src/protocols/tarball-fetcher.md>
+
+use anyhow::{Context, Result};
+use reqwest::header::HeaderName;
+use serde::{Deserialize, Serialize};
+use url::Url;
+
+use crate::*;
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
+pub struct TarballPin {
+    /// URL provided as user input.
+    ///
+    /// Will get moved into `LockedTarball` once the Pin has been updated at
+    /// least once. This is a little bit of a hack but we need to be able to
+    /// update the `url` from within `TarballPin::update` so we have to store it
+    /// within `LockedTarball` which doesn't exist yet the first time `update`
+    /// is called.
+    #[serde(skip)]
+    pub url: Option<Url>,
+}
+
+impl diff::Diff for TarballPin {
+    fn properties(&self) -> Vec<(String, String)> {
+        vec![]
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
+pub struct LockedTarball {
+    /// If the given URL supports the Lockable Tarball Protocol we use the
+    /// locked one for `url` and we store the original used for updating here
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub lockable_url: Option<Url>,
+    /// Either the URL provided by user or an immutable URL returned in Link header
+    pub url: Url,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub revision: Option<String>,
```

Comment:
Given that we're not implementing the full Flake URL specification and are using a plain tarball fetcher, maybe let's skip these for now altogether.

Reply:
oki, i've removed the whole bit